server/src/main/java/org/elasticsearch/common/blobstore/RetryingInputStream.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.blobstore;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.repositories.blobstore.RequestedRangeNotSatisfiedException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.core.Strings.format;
+
+public abstract class RetryingInputStream<V> extends InputStream {
+
+    private static final Logger logger = LogManager.getLogger(RetryingInputStream.class);
+
+    public static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private final BlobStoreServices<V> blobStoreServices;
+    private final OperationPurpose purpose;
+    private final long start;
+    private final long end;
+    private final List<Exception> failures;
+
+    protected InputStreamAtVersion<V> currentStream;
+    private long currentStreamFirstOffset;
```

Comment:
Is it possible to manage `currentStreamProgress` within `currentStream`?

Reply:
I think it would mean we'd need to re-implement that tracking for each CSP? I guess we could have an abstract class for the "current-stream" input stream and implement it there, but that just adds more inheritance to trip over.

We use it to determine whether we've made "meaningful progress" which I had considered a cross-cutting concern for the purpose of this re-use.

Reply:
Cool, I added that in b30ca9e6d54af7844940cba036cdb240d34868ad. I think it looks nicer.